src/zelos/feeds.py

```diff
@@ -0,0 +1,226 @@
+    has been called, specify '--inst_feed=syscall=recv' on the command
+    line. For a script, add 'inst_feed="syscall=recv"' as a keyword
+    argument in the Zelos constructor.
+
+For more information on what options are available for configuring feeds
+look at the zelos.zml module.
+"""
+
+
+import functools
+import logging
+
+from enum import IntEnum
+from typing import Callable, Dict
+
+from zelos.hooks import HookManager, HookType
+from zelos.zml import ZmlParser
+
+
+class FeedLevel(IntEnum):
+    NONE = 0
+    SYSCALL = 1
+    API = 2
+    INST = 3
+
+
+class FeedHandle:
+    """
+    Returned when subscribing to a feed. Used for unsubscribing to a
+    feed.
+    """
+
+    def __init__(self, feed_level, num):
+        self._feed_level = feed_level
+        self._num = num
+
+
+class FeedManager:
+    """
+    Handles feed subscribers as well as the feed level.
+
+    Subscription is handled by passing a callback to the subscribe_to_*
+    functions.
+    """
+
+    def __init__(
+        self, config, zml_parser: ZmlParser, hook_manager: HookManager
+    ):
+        self._hook_manager = hook_manager
+        self.logger = logging.getLogger(__name__)
+
+        # _inst_hook_info is only present when the inst feed is active.
+        self._inst_hook_info = None
+        # The syscall hook must be registered later
+        self._syscall_hook_info = None
+
+        self._feed_level = FeedLevel.NONE
+
+        self._handle_num = 0
+        self._subscribers: Dict[FeedLevel, Dict[int, Callable]] = {
+            FeedLevel.SYSCALL: {},
+            FeedLevel.API: {},
+            FeedLevel.INST: {},
+        }
+
+        # Default to feed level SYSCALL, unless plans for the syscall
+        # level feed were already specified.
+        if len(config.syscall_feed) == 0:
+            self.set_feed_level(FeedLevel.SYSCALL)
+
+        # For initial setup, we want to respect the highest level set if
+        # multiple are set to trigger immediately. We run them from
+        # lowest to highest to achieve this goal
+        for zml_string in config.stop_feed:
+            zml_parser.trigger_on_zml(
+                functools.partial(self.set_feed_level, FeedLevel.NONE),
+                zml_string,
+            )
+
+        for zml_string in config.api_feed:
+            zml_parser.trigger_on_zml(
+                functools.partial(self.set_feed_level, FeedLevel.API),
+                zml_string,
+            )
+
+        for zml_string in config.syscall_feed:
+            zml_parser.trigger_on_zml(
+                functools.partial(self.set_feed_level, FeedLevel.SYSCALL),
+                zml_string,
+            )
+
+        for zml_string in config.inst_feed:
+            zml_parser.trigger_on_zml(
+                functools.partial(self.set_feed_level, FeedLevel.INST),
+                zml_string,
+            )
+
+        # TODO: Turning on the syscall feed from a syscall hook skips
```

Comment:
Instead of a hook, you could invoke this feed function directly from the base syscall manager (BEFORE). This would also help reduce the total code path executed each syscall. Minimizing required hooks is critical for performance.

Reply:
Moved to syscall manager base. On the surface, I can't see how this reduces the total code path executed. We should incorporate the frequent syscall test to check what the slowdown would be.